### PR TITLE
Sdk-776 MC Block reference delay

### PR DESCRIPTION
**Pull request by VolodymyrBabaiev**
Introduced MC block reference delay. It should be set at the application layer.
Updated python tests, introduced tests for checking pending certificates in non-ceasing sidechains and interaction forgers with different MC block reference delay.

### CODE REVIEW

**Review by paolocappelletti — CHANGES_REQUESTED** [review]

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/io/horizen/forge/AbstractForgeMessageBuilder.scala [hunk 1]
```diff
@@ -186,19 +186,21 @@ abstract class AbstractForgeMessageBuilder[
             throw ex
       }
 
+    newHeaderHashes = if(newHeaderHashes.size >= 1) newHeaderHashes.take(newHeaderHashes.size - params.mcBlockRefDelay) else Seq()
```
[the remark this answers is not in the document]
Comment (paolocappelletti) [reply sdk/src/main/scala/io/horizen/forge/AbstractForgeMessageBuilder.scala:189]: newHeaderHashes.size must also be > params.mcBlockRefDelay

**Review by paolocappelletti — CHANGES_REQUESTED** [review]
should be nice to add also at least one test with a delay > 1 (I would try 6)

File: sdk/src/main/scala/io/horizen/AbstractSidechainApp.scala [hunk 2]
```diff
@@ -67,6 +68,7 @@ abstract class AbstractSidechainApp
   private val closableResourceList = mutable.ListBuffer[AutoCloseable]()
   protected val sidechainTransactionsCompanion: DynamicTypedSerializer[TX, TransactionSerializer[TX]]
   protected val terminationTimeout: FiniteDuration = Duration(30, TimeUnit.SECONDS)
+  protected val maxMcBlockRefDelay = 10
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply sdk/src/main/scala/io/horizen/AbstractSidechainApp.scala:71]: This value was discussed with Paolo and mentioned in design document https://docs.google.com/document/d/1pZDKDffFemEliOBlBGjmoFgL3Eu98bmbQUuNzVM1hhc/edit?usp=sharing

File: sdk/src/main/scala/io/horizen/forge/AbstractForgeMessageBuilder.scala [hunk 3]
```diff
@@ -186,19 +186,21 @@ abstract class AbstractForgeMessageBuilder[
...
         withdrawalEpochMcBlocksLeft = params.withdrawalEpochLength
 
-      // to not to include mcblock references data from different withdrawal epochs
-      val maxReferenceDataNumber: Int = withdrawalEpochMcBlocksLeft
-
       val missedMainchainReferenceDataHeaderHashes: Seq[MainchainHeaderHash] = history.missedMainchainReferenceDataHeaderHashes
       val nextMainchainReferenceDataHeaderHashes: Seq[MainchainHeaderHash] = missedMainchainReferenceDataHeaderHashes ++ newHeaderHashes
 
+      // to not to include mcblock references data from different withdrawal epochs
+      val maxReferenceDataNumber: Int = Math.min(withdrawalEpochMcBlocksLeft, nextMainchainReferenceDataHeaderHashes.size)
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply sdk/src/main/scala/io/horizen/forge/AbstractForgeMessageBuilder.scala:202]: There are 2 limitations of how many blocks we can include - one is MC block delay(we don't want to include some blocks to make delay), the other one is we don't include to one SC block reference of mainchain block form different epochs.

File: examples/utxo/simpleapp/src/main/java/io/horizen/examples/SimpleApp.java [hunk 4]
```diff
@@ -24,7 +24,17 @@ public static void main(String[] args) {
         }
         String settingsFileName = args[0];
 
-        Injector injector = Guice.createInjector(new SimpleAppModule(settingsFileName));
+        int mcBlockReferenceDelay = 0;
+        try {
```
[the remark this answers is not in the document]
Comment (VolodymyrBabaiev) [reply examples/utxo/simpleapp/src/main/java/io/horizen/examples/SimpleApp.java:28]: Actually we don't want this information to be passed by command line or set in config. It must set on application layer and be same for every node in the Sidechain. We have to pass it in command line only for testing purpose.

**Review by paolocappelletti — APPROVED** [review]

**Review by MarcoOl94 — APPROVED** [review]

**Pull request merged**